libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java

```diff
@@ -0,0 +1,288 @@
+        return syncManager;
+    }
+
+    /**
+     * Fetches layout data for the specified object type and layout type using the specified
+     * mode and triggers the supplied callback once complete.
+     *
+     * @param objectType Object type.
+     * @param layoutType Layout type. Defaults to "Full" if null is passed in.
+     * @param mode Fetch mode. See {@link Mode} for available modes.
+     * @param syncCallback Layout sync callback.
+     */
+    public void fetchLayout(String objectType, String layoutType, Mode mode,
+                            LayoutSyncCallback syncCallback) {
+        switch (mode) {
+            case CACHE_ONLY:
+                fetchFromCache(objectType, layoutType, syncCallback, false);
+                break;
+            case CACHE_FIRST:
+                fetchFromCache(objectType, layoutType, syncCallback, true);
+                break;
+            case SERVER_FIRST:
+                fetchFromServer(objectType, layoutType, syncCallback);
+                break;
+        }
+    }
+
+    private LayoutSyncManager(SmartStore smartStore, SyncManager syncManager) {
+        this.smartStore = smartStore;
+        this.syncManager = syncManager;
+        initializeSoup();
+    }
+
+    private void fetchFromServer(final String objectType, final String layoutType,
+                                 final LayoutSyncCallback syncCallback) {
+        final SyncDownTarget target = new LayoutSyncDownTarget(objectType, layoutType);
+        final SyncOptions options = SyncOptions.optionsForSyncDown(SyncState.MergeMode.OVERWRITE);
+        try {
+            syncManager.syncDown(target, options, SOUP_NAME, new SyncManager.SyncUpdateCallback() {
+
+                @Override
+                public void onUpdate(SyncState sync) {
+                    if (SyncState.Status.DONE.equals(sync.getStatus())) {
+                        fetchFromCache(objectType, layoutType, syncCallback, false);
+                    }
+                }
+            });
+        } catch (Exception e) {
+            SmartSyncLogger.e(TAG, "Exception occurred while reading layout data from the server", e);
+        }
+    }
+
+    private void fetchFromCache(String objectType, String layoutType,
+                                LayoutSyncCallback syncCallback, boolean fallbackOnServer) {
+        final QuerySpec querySpec = QuerySpec.buildSmartQuerySpec(String.format(QUERY,
+                objectType, layoutType), 1);
+        try {
+            final JSONArray results = smartStore.query(querySpec, 0);
+            if (results == null || results.length() == 0) {
+                if (fallbackOnServer) {
+                    fetchFromServer(objectType, layoutType, syncCallback);
+                } else {
+                    onSyncComplete(objectType, syncCallback, null);
+                }
+            } else {
+                final JSONArray resultArr = results.optJSONArray(0);
+                if (resultArr == null || resultArr.length() == 0) {
```

Comment:
When would that happen? And shouldn't we then check fallbackOnServer.
In other words should those conditions move to line 228?

Reply:
Nope - this would probably happen when the server returned an error of some kind, or the object doesn't have a layout or something like that (I don't know if that's a real possibility).

Reply:
Not sure I follow. If there is an error during sync down, nothing should be saved to the table so the first if should be entered.

Reply:
True. I guess this condition should never occur or would occur only if the response was a valid empty payload. Is that even possible?

Reply:
Okay - will remove it.